Cookbook/Proposals/FeatureSwitchesRedesign.md

```diff
@@ -0,0 +1,218 @@
+# Feature switches redesign
+
+* Author(s): Ilya Puchka
+* Review Manager: TBD
+
+## Introduction
+
+This proposal aims to define a unified approach to defining and accessing feature switches comming from various sources in our code base.
+
+## Motivation
+
+Even though feature switches in essens is a trivial thing our current apporach to them suffers from multiple issues of difference importance.
+
+### Root plist and default values
+
+Currently we provide a way to override any feature switch through toggles in the Settings system app. For that we need to include `Root.plist` file in the app bundle. Values of these switches are then written to the user defaults which we use as a source of values. But they are written to the user defaults only after they are interacted at least once. That said if overrides are enabled but you never change the value nothing will be written to the user defaults. This is not affected by default value set in the `Root.plist`, it only affects the intial position of the switch. How we solve it now is by reading these default values from plist and storing them in user defaults when application starts before we read these values (see `registerSettingsDefaults` in `Settings.swift`). Not just it is defined on the wrong type (`Settings` type is used to store some local user state data like if they completed some onboarding, not feature switches and this is not something that we allow to override with `Root.plist` AFAIK), but also this way changing default value in the `Root.plist` will affect all our targets as they all share this file, which makes us to create a second file just for our primary target. This makes it very confusing for developers to understand what they need to do to enable their feature for release so that it does not affect other targets if that's the requirement.
+
+Without this syncing we would have to deal with optionals for all the switches, falling back to their defaults, defined in code. That also highlights another issue with `Root.plist`, that we need to repeat default value in the code and in the plist and make sure they allign to avoid confusion.
+
+Another issue caused by using `Root.plist` and overrides in this way is when we move from Local feature switch used during development to Remote feature switch used on a final stage of the feautre rollout we have to remember to move the plist entry for this switch from one section to another. Otherwise there will be no way to override it during testing.
```

Comment:
```suggestion
Another issue caused by using `Root.plist` and overrides in this way is when we move from Local feature switch used during development to Remote feature switch used on a final stage of the feature rollout we have to remember to move the plist entry for this switch from one section to another. Otherwise there will be no way to override it during testing.
```